File: app/models/planning/execution_plan.rb
# frozen_string_literal: true

module Planning
  # Represents a top-level execution plan with milestones and metadata.
  # This is the primary result object returned by the PlanAgentWorker.
  #
  # @example Creating an execution plan
  #   plan = Planning::ExecutionPlan.new(
  #     goal: "Create a PlanAgentWorker",
  #     milestones: [milestone1, milestone2],
  #     constraints: ["Must follow OOP patterns"],
  #     assumptions: ["Rails environment available"]
  #   )
  class ExecutionPlan
    attr_reader :goal, :milestones, :created_at, :metadata,
                :constraints, :assumptions, :risks

    # @param goal [String] The goal this plan achieves
    # @param milestones [Array<PlanMilestone>] Array of milestones
    # @param created_at [String, nil] ISO8601 timestamp (generated if not provided)
    # @param metadata [Hash] Additional metadata
    # @param constraints [Array<String>, nil] Optional constraints to follow
    # @param assumptions [Array<String>, nil] Optional assumptions made
    # @param risks [Array<String>, nil] Optional identified risks
    def initialize(goal:, milestones:, created_at: nil, metadata: {},
                   constraints: nil, assumptions: nil, risks: nil)
      validate_required!(goal, milestones, metadata)
      validate_optional!(constraints, assumptions, risks)

      @goal = goal
      @milestones = Array(milestones)
      @created_at = created_at || Time.now.utc.iso8601
      @metadata = metadata
      @constraints = constraints
      @assumptions = assumptions
      @risks = risks
    end

    # Add a milestone to the plan
    # @param milestone [PlanMilestone] Milestone to add
    # @return [PlanMilestone] The added milestone
    def add_milestone(milestone)
      unless milestone.is_a?(PlanMilestone)
        raise ArgumentError, "milestone must be a PlanMilestone, got #{milestone.class}"
      end

      @milestones << milestone
      milestone
    end

    # Count of milestones in the plan
    # @return [Integer]
    def milestone_count
      @milestones.size
    end

    # Total count of steps across all milestones
    # @return [Integer]
    def step_count
      @milestones.sum(&:step_count)
    end

    # Serialize to hash for persistence (recursive with milestones)
    # @return [Hash] Hash representation of the plan
    def to_h
      {
        goal: @goal,
        milestones: @milestones.map(&:to_h),
        created_at: @created_at,
        metadata: @metadata,
        constraints: @constraints,
        assumptions: @assumptions,
        risks: @risks
      }
    end

    # Reconstruct an ExecutionPlan from a hash
    # @param hash [Hash] Hash containing plan data
    # @return [Planning::ExecutionPlan] Reconstructed plan
    def self.from_h(hash)
      raise ArgumentError, "hash must be a Hash, got #{hash.class}" unless hash.is_a?(Hash)

      # Handle both symbol and string keys
      goal = hash[:goal] || hash["goal"]
      milestones_data = hash[:milestones] || hash["milestones"] || []
      created_at = hash[:created_at] || hash["created_at"]
      metadata = hash[:metadata] || hash["metadata"] || {}
      constraints = hash[:constraints] || hash["constraints"]
      assumptions = hash[:assumptions] || hash["assumptions"]
      risks = hash[:risks] || hash["risks"]

      # Reconstruct milestones
      milestones = milestones_data.map { |milestone_hash| PlanMilestone.from_h(milestone_hash) }

      new(
        goal: goal,
        milestones: milestones,
        created_at: created_at,
        metadata: metadata,
        constraints: constraints,
        assumptions: assumptions,
        risks: risks
      )
    end

    private

    def validate_required!(goal, milestones, metadata)
      # Validate goal
      unless goal.is_a?(String) && !goal.strip.empty?
        raise ArgumentError, "goal must be a non-empty String, got #{goal.class}. " \
                             "Example: 'Create a PlanAgentWorker'"
      end

      # Validate milestones
      raise ArgumentError, "milestones must be an Array, got #{milestones.class}" unless milestones.is_a?(Array)
      if milestones.any? { |m| !m.is_a?(PlanMilestone) }
        raise ArgumentError, "all milestones must be PlanMilestone instances"
      end

      # Validate metadata
      raise ArgumentError, "metadata must be a Hash, got #{metadata.class}" unless metadata.is_a?(Hash)
    end

    def validate_optional!(constraints, assumptions, risks)
      # Validate constraints (if provided)
      if constraints && !constraints.is_a?(Array)
        raise ArgumentError, "constraints must be an Array, got #{constraints.class}"
      end

      # Validate assumptions (if provided)
      if assumptions && !assumptions.is_a?(Array)
        raise ArgumentError, "assumptions must be an Array, got #{assumptions.class}"
      end

      # Validate risks (if provided)
      if risks && !risks.is_a?(Array)
        raise ArgumentError, "risks must be an Array, got #{risks.class}"
      end
    end
  end
end

File: frontend/src/models/ExecutionPlan.js
// ExecutionPlan domain model - represents a Daedalus execution plan
// Follows strict OOP patterns from oop-patterns.md

export class ExecutionPlan {
  constructor({
    goal,
    constraints,
    assumptions,
    risks,
    milestones,
    outputPaths,
    analysisSummary,
    metadata,
  }) {
    // Validate required parameters
    if (!goal || typeof goal !== "string") {
      throw new Error("goal must be a non-empty string");
    }
    if (!Array.isArray(constraints)) {
      throw new Error("constraints must be an array");
    }
    if (!Array.isArray(assumptions)) {
      throw new Error("assumptions must be an array");
    }
    if (!Array.isArray(risks)) {
      throw new Error("risks must be an array");
    }
    if (!Array.isArray(milestones)) {
      throw new Error("milestones must be an array");
    }
    if (!outputPaths || typeof outputPaths !== "object") {
      throw new Error("outputPaths must be an object");
    }
    if (!analysisSummary || typeof analysisSummary !== "object") {
      throw new Error("analysisSummary must be an object");
    }
    if (!metadata || typeof metadata !== "object") {
      throw new Error("metadata must be an object");
    }

    this._goal = goal;
    this._constraints = Object.freeze([...constraints]);
    this._assumptions = Object.freeze([...assumptions]);
    this._risks = Object.freeze([...risks]);
    this._milestones = Object.freeze(
      milestones.map((m) => Object.freeze({ ...m }))
    );
    this._outputPaths = Object.freeze({ ...outputPaths });
    this._analysisSummary = Object.freeze({ ...analysisSummary });
    this._metadata = Object.freeze({ ...metadata });

    Object.freeze(this);
  }

  get goal() {
    return this._goal;
  }

  get constraints() {
    return this._constraints;
  }

  get assumptions() {
    return this._assumptions;
  }

  get risks() {
    return this._risks;
  }

  get milestones() {
    return this._milestones;
  }

  get outputPaths() {
    return this._outputPaths;
  }

  get analysisSummary() {
    return this._analysisSummary;
  }

  get metadata() {
    return this._metadata;
  }

  // Get milestone by index
  getMilestone(index) {
    return this._milestones[index] || null;
  }

  // Get total milestone count
  getMilestoneCount() {
    return this._milestones.length;
  }

  // Get total step count across all milestones
  getStepCount() {
    return this._milestones.reduce(
      (total, milestone) => total + (milestone.steps?.length || 0),
      0
    );
  }

  // Serialize to JSON
  toJSON() {
    return {
      goal: this._goal,
      constraints: this._constraints,
      assumptions: this._assumptions,
      risks: this._risks,
      milestones: this._milestones,
      output_paths: this._outputPaths,
      analysis_summary: this._analysisSummary,
      metadata: this._metadata,
    };
  }

  // Deserialize from JSON (from API response)
  static fromJSON(json) {
    if (!json || typeof json !== "object") {
      throw new Error("JSON must be an object");
    }

    // Handle both camelCase (frontend) and snake_case (backend) keys
    const executionPlan = json.execution_plan || json;

    return new ExecutionPlan({
      goal: executionPlan.goal,
      constraints: executionPlan.constraints || [],
      assumptions: executionPlan.assumptions || [],
      risks: executionPlan.risks || [],
      milestones: executionPlan.milestones || [],
      outputPaths: json.output_paths || executionPlan.output_paths || {},
      analysisSummary:
        json.analysis_summary || executionPlan.analysis_summary || {},
      metadata: json.metadata || {},
    });
  }

  // Create from Daedalus API response
  static fromDaedalusResponse(response) {
    if (!response || !response.result) {
      throw new Error("Invalid Daedalus response");
    }

    return ExecutionPlan.fromJSON({
      execution_plan: response.result.execution_plan,
      output_paths: response.result.output_paths,
      analysis_summary: response.result.analysis_summary,
      metadata: response.metadata,
    });
  }
}


